### ISSUE TITLE
[$250] Can't create hyperlink with different protocol other than HTTP or HTTPS reported by @kerupuksambel

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#12290]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open [staging.new.expensify.com](http://staging.new.expensify.com/)
2. Send this HTML snippet to someone : [ftp://google.com/](ftp://google.com/)

## Expected Result:
The FTP link should be rendered as hyperlink (e.g. ftp://google.com/)

## Actual Result:
The markdown doesn't rendered and it's only put as it is ([ftp://google.com/](ftp://google.com/))

## Workaround:
unknown

## Platform:

Where is this issue occurring?

- Web

**Version Number:** 1.2.21-4

**Reproducible in staging?:** y

**Reproducible in production?:** y

**Email or phone of affected tester (no customers):** 

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/198912565-9e296d72-0006-4cbe-9ab0-1e558826ae09.mp4

<img width="960" alt="ftp" src="https://user-images.githubusercontent.com/43996225/198912569-8fcdf983-7000-4a2c-a06f-4bae14e124a6.png">

**Expensify/Expensify Issue URL:** 

**Issue reported by:** @kerupuksambel

**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1666996403875309

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by lschurr** [comment]
I was able to reproduce. Making sure there isn't another open issue for this. https://expensify.slack.com/archives/C01GTK53T8Q/p1667236678452169?thread_ts=1666996403.875309&cid=C01GTK53T8Q

**Comment by lschurr** [comment]
Hey @marcaaron - could you just double check this and confirm this is different from other reports? Thread [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1667236678452169?thread_ts=1666996403.875309&cid=C01GTK53T8Q).

**Comment by marcaaron** [comment]
Looks new to me and agree it can be it's own bug I think. 

Separately, whoever fixes our link parser should also add a unit test for this.

**Comment by parasharrajat** [comment]
Looks like we never supported protocols other than HTTP.
![image](https://user-images.githubusercontent.com/24370807/199285593-f15cc49e-507c-4b7d-9d52-646903927d9e.png)

**Comment by lschurr** [comment]
Posted job:
https://www.upwork.com/ab/applicants/1587482663583879168/job-details
https://www.upwork.com/jobs/~01e8c596a397abd681

**Comment by kerupuksambel** [comment]
Sorry for joining the discussion suddenly, but I think if we preferred some kind of hard-coded list about what to add, I think we can refer to [IANA's URI scheme](https://en.wikipedia.org/wiki/List_of_URI_schemes#Official_IANA-registered_schemes), along with additional [unofficial, but popular URI scheme](https://en.wikipedia.org/wiki/List_of_URI_schemes#Unofficial_but_common_URI_schemes) that could be added. Some suggestions maybe including schemes like `tel` and `mailto` for contacting phone number and email address respectively, `zoommtg` for launching Zoom client, `slack` for launching Slack client, `msteams` for launching Microsoft Teams client, `git` for Git repository, etc. 

However, if the list approach seems to be too long or not flexible, I think we could use some regex wildcards to get the valid schemes to be allowed. The simple example I could think of is `(([a-z0-9\-\+\.])+)`. (Reference from [URI syntax reference in Wikipedia](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Syntax)). However, it seems that there are several edge cases that should be covered such as `tel` and `mailto` which doesn't have double slash at its scheme.

**Comment by parasharrajat** [comment]
Good question, @lschurr What protocols are we expecting to be supported? This issue states `ftp`. Are there more?

**Comment by lschurr** [comment]
@marcaaron - are you able to weigh in on these questions?

**Comment by marcaaron** [comment]
imo we should not have an allow list for specific scheme. In the world of deep links any text based scheme should be potentially valid.

**Comment by marcaaron** [comment]
seems like @kerupuksambel has a good grasp of this subject. Will you consider putting a proposal together so we get you working on this?

**Comment by kerupuksambel** [comment]
@marcaaron Sure thing! 

# Proposal

The problem here, as mentioned by @parasharrajat, is that the URI parser only matched the one with HTTP/HTTPS scheme. Therefore, my proposal is to check all the valid URI scheme.

My first proposal is going to change this 
```javascript
const URL_WEBSITE_REGEX = `(https?:\\/\\/)?((?:www\\.)?[-a-z0-9]+?\\.)+(?:${TLD_REGEX})(?:\\:\\d{2,4}|\\b|(?=_))`;
```
in https://github.com/Expensify/expensify-common/blob/a71aad37aaba2b43b3a2d2f4be8052b025b265d3/lib/ExpensiMark.js#L5 into this (slashes in Regex was escaped due to the nature of Javascript string)
```javascript
const URL_WEBSITE_REGEX = `(([a-z0-9\\+\\-\\.])+:\\/?\\/?)?((?:www\\.)?[-a-z0-9@]+?\\.)+(?:com|net|id|org)(?:\\:\\d{2,4}|\\b|(?=_))`
```
Checking with regex101.com, the new regex have some satisfying result on some test cases that wasn't covered by the old regex.
![image](https://user-images.githubusercontent.com/24503760/199386677-d0939ceb-b492-4531-b2b9-c8cfa3394d0d.png)

**Comment by parasharrajat** [comment]
I am reviewing it.

**Comment by parasharrajat** [comment]
This suggestion literally accepts every protocol which I think is fine. But this is not complete. 

It does not handle link conversion. If you pass through the parser `ftp://google.com/` with your solution. 

You will get `ftp://google.com/` as the text of the link and `https://ftp://google.com/` as URL which is incorrect.

**Comment by mallenexpensify** [comment]
Bumped back to daily per our new BugZero process, [internal Slack thread with deets](https://expensify.slack.com/archives/C01SKUP7QR0/p1667064948124109)

**Comment by kerupuksambel** [comment]
@parasharrajat Really sorry for my late reply, I was sick a few days back then. Also, I completed the changes so it should be working as proposed. Therefore, if possible, I would like to revise my proposal.

## Revised Proposal
Upon my further analysis to fix this bug, I found out several things, such as possibility to do XSS by `javascript` scheme. My take on that problem is to filter everything with `javascript` scheme (for example `javascript:alert(1)`).

Also, as a problem pointed by @parasharrajat, now the fixed proposal also should create the hyperlink with correct scheme as well.

Therefore, this is my proposal for the change of the code in `lib/ExpensiMark.js` in [expensify-common](https://github.com/Expensify/expensify-common) repository.

```diff
diff --git a/lib/ExpensiMark.js b/lib/ExpensiMark.js
index 1b7c39d..6eb1ffe 100644
--- a/lib/ExpensiMark.js
+++ b/lib/ExpensiMark.js
@@ -2,7 +2,7 @@ import _ from 'underscore';
 import Str from './str';
 import TLD_REGEX from './tlds';
 
-const URL_WEBSITE_REGEX = `(https?:\\/\\/)?((?:www\\.)?[-a-z0-9]+?\\.)+(?:${TLD_REGEX})(?:\\:\\d{2,4}|\\b|(?=_))`;
+const URL_WEBSITE_REGEX = `(([a-z0-9\\+\\-\\.])+:\\/?\\/?)?((?:www\\.)?[-a-z0-9@]+?\\.)+(?:${TLD_REGEX})(?:\\:\\d{2,4}|\\b|(?=_))`
 const addEscapedChar = reg => `(?:${reg}|&(?:amp|quot|#x27);)`;
 const URL_PATH_REGEX = `(?:${addEscapedChar('[.,=(+$!*]')}?\\/${addEscapedChar('[-\\w$@.+!*:(),=%~]')}*${addEscapedChar('[-\\w~@:%)]')}|\\/)*`;
 const URL_PARAM_REGEX = `(?:\\?${addEscapedChar('[-\\w$@.+!*()\\/,=%{}:;\\[\\]\\|_]')}+)?`;
@@ -81,13 +81,14 @@ export default class ExpensiMark {
                 },
 
                 // We use a function here to check if there is already a https:// on the link.
+                // Also, we prevented XSS by filter whether the protocol is javascript or not
                 // If there is not, we force the link to be absolute by prepending '//' to the target.
                 replacement: (match, g1, g2, g3) => {
                     if (!g1.trim()) {
                         return match;
                     }
 
-                    return `<a href="${g3 && g3.includes('http') ? '' : 'https://'}${g2}" target="_blank" rel="noreferrer noopener">${g1}</a>`;
+                    return `<a href="${g3 !== undefined && !g3.toLowerCase().includes("javascript") ? '' : 'https://'}${g2}" target="_blank" rel="noreferrer noopener">${g1}</a>`;
                 },
             },
 
@@ -117,9 +118,12 @@ export default class ExpensiMark {
                 },
 
                 // We use a function here to check if there is already a https:// on the link.
+                // Also, we prevented XSS by filter whether the protocol is javascript or not
                 // If there is not, we force the link to be absolute by prepending '//' to the target.
+
+
                 replacement: (match, g1, g2, g3) => (
-                    `${g1}<a href="${g3 && g3.includes('http') ? '' : 'https://'}${g2}" target="_blank" rel="noreferrer noopener">${g2}</a>${g1}`
+                    `${g1}<a href="${g3 !== undefined && !g3.toLowerCase().includes("javascript") ? '' : 'https://'}${g2}" target="_blank" rel="noreferrer noopener">${g2}</a>${g1}`
                 ),
             },
             {

```

**Comment by parasharrajat** [comment]
Thank you for the updated proposal. Let me discuss this internally to get some details on https://github.com/Expensify/App/issues/12290#issuecomment-1299216934

**Comment by parasharrajat** [comment]
Asked here https://expensify.slack.com/archives/C01GTK53T8Q/p1667828343990809.

**Comment by lschurr** [comment]
Seems like we're going to move forward, right @parasharrajat?

**Comment by parasharrajat** [comment]
I need to test a few things before we can. Let me do that first.

**Comment by parasharrajat** [comment]
@kerupuksambel tests are failing. Can you please test your proposal and fix it so that it does not break old tests? Also, you will have to add more unit tests to test the new functionality as part of this task. 


```js
  const URL_WEBSITE_REGEX = `(([a-z0-9\\+\\-\\.])+:\\/?\\/?)?((?:www\\.)?[-a-z0-9@]+?\\.)+
 																				 ^
```
Why the @ was added to this regex here?

**Comment by kerupuksambel** [comment]
Hi @parasharrajat, first really apologize again for my slow response towards the issue. And yes, so far the test without the "@" here would pass the test. The "@" here supposed to handle email protocol, however upon better inspection towards the code, I realized the email has been handled on the code above, therefore it should be no longer needed on the code. Therefore, the revised regex should be :

```javascript
const URL_WEBSITE_REGEX = `(([a-z0-9\\+\\-\\.])+:\\/?\\/?)?((?:www\\.)?[-a-z0-9]+?\\.)+(?:${TLD_REGEX})(?:\\:\\d{2,4}|\\b|(?=_))`
```

And the regex passed the test (with added test units below)
![image](https://user-images.githubusercontent.com/24503760/201463825-6988de71-c27f-42f6-b317-8370198d2e12.png)


As for test case, I propose this unit tests for testing the functionality of the code.

```diff
diff --git a/__tests__/ExpensiMark-HTML-test.js b/__tests__/ExpensiMark-HTML-test.js
index e1d1b2f..5895a74 100644
--- a/__tests__/ExpensiMark-HTML-test.js
+++ b/__tests__/ExpensiMark-HTML-test.js
@@ -337,6 +337,25 @@ test('Test markdown style email link with various styles', () => {
     expect(parser.replace(testString)).toBe(resultString);
 });
 
+
+test('Test URI with different schemes, while also filtering dangerous schemes (javascript)', () => {
+    const testString = 'This is [FTP link](ftp://expensify.com), '
+        + '[zoom-mtg link](zoom-mtg://expensify.com), '
+        + '[arbitrary scheme link](abc-def+12.3://expensify.com), '
+        + '[scheme with mixed case](FtP://expensify.com), '
+        + '[but still ensure no http still parsed as usual](expensify.com), '
+        + '[and javascript scheme should be filtered for better security.](javascript:console.com)';
+    
+    const resultString = 'This is <a href="ftp://expensify.com" target="_blank" rel="noreferrer noopener">FTP link</a>, '
+        + '<a href="zoom-mtg://expensify.com" target="_blank" rel="noreferrer noopener">zoom-mtg link</a>, '
+        + '<a href="abc-def+12.3://expensify.com" target="_blank" rel="noreferrer noopener">arbitrary scheme link</a>, '
+        + '<a href="FtP://expensify.com" target="_blank" rel="noreferrer noopener">scheme with mixed case</a>, '
+        + '<a href="https://expensify.com" target="_blank" rel="noreferrer noopener">but still ensure no http still parsed as usual</a>, '
+        + '<a href="https://javascript:console.com" target="_blank" rel="noreferrer noopener">and javascript scheme should be filtered for better security.</a>';
+    
+    expect(parser.replace(testString)).toBe(resultString);    
+})
+
 test('Test general email link with various styles', () => {
     const testString = 'Go to concierge@expensify.com '
         + 'no-concierge@expensify.com '
```

**Comment by parasharrajat** [comment]
Yup. Sure.

**Comment by parasharrajat** [comment]
Ok so I tested it. It is looking good.  

the _javascript: scheme part_ . let's say we the parser is parsing
```
javascript:console.com
```
Your solution will give us 
```
https://javascript:console.com
```
Not sure what should be the output here. Does this look correct? For me, it is fine as it is a https url. 

--- 
@kerupuksambel 's [proposal](https://github.com/Expensify/App/issues/12290#issuecomment-1304739804) + url regex from https://github.com/Expensify/App/issues/12290#issuecomment-1312401880 looks good to me. 

I am not security expert but I will suggest someone from internal team to review this. 

cc: @marcaaron 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by marcaaron** [comment]
>javascript scheme should be filtered for better security

My 2 cents I really don't think this is a requirement for this issue

**Comment by lschurr** [comment]
@parasharrajat are we moving forward with @kerupuksambel's proposal?

**Comment by parasharrajat** [comment]
Currently, the staging server is doing the first one, excluding `javascript:` from autolinking.

**Comment by parasharrajat** [comment]
I would say that we do not update the existing parsing logic for `javascript:console.com` in this issue and just go over the protocol parsing. But if you say that we handle this case as well, I am fine too.

Please let me know and I will expect the same in proposals.

**Comment by marcaaron** [comment]
No worries and thank you @parasharrajat. If that's the current behavior then we can create a separate issue or get wider feedback on whether it is something that really needs to be fixed.

Definitely no way should `javascript:console.com` become `https://javascript:console.com`, but I'm sure we agree there.

**Comment by parasharrajat** [comment]
> Definitely no way should javascript:console.com become https://javascript:console.com, but I'm sure we agree there.

Yup. Agreed.

**Comment by lschurr** [comment]
Cool, are we moving forward then or do we need an update to the proposal from @kerupuksambel?

**Comment by marcaaron** [comment]
We need an update (or at least confirmation that the behavior will be fixed in the PR review).

**Comment by parasharrajat** [comment]
Bump @kerupuksambel

**Comment by puneetlath** [comment]
@marcaaron this is one of the oldest issues in the /App repo. To help us clear out the large backlog of bugs, can you:  
- Decide whether any proposals currently meet our guidelines and can be approved as-is 
- For any that can't, please take this issue internal and treat it as one of your highest priorities 
- If you have any questions, don't hesitate to start a discussion in #bug-zero

**Comment by kerupuksambel** [comment]
Firstly, thank you for the initiative to removing the `javascript` filter. Really apologize for not engaging into the discussion since this week I've been sick. 

My consideration for filtering the javascript scheme is because there are several payload that could be used for executing cross-site scripting (XSS) attack by using javascript scheme (for example https://www.bugbountyhunter.com/hackevents/report?id=121 and https://cheatsheetseries.owasp.org/cheatsheets/XSS_Filter_Evasion_Cheat_Sheet.html). 

With the possible exploit could possibly be sent and working, it's possible for an attacker to receive important data like the victim's credentials if the victim clicked the link. However, I can't find the working javascript in time, so I made the `javascript:console.com` as an example for the javascript scheme payload.

Adding the `https://` on top of `javascript:` scheme would be disabling those vulnerability, since the scheme now is HTTPS and not javascript anymore, the browser would treat them like a regular HTTPS link when clicked, and not a javascript command. 

Here's an example for the structure of an unfiltered `javascript` scheme compared to the filtered one.
![image](https://user-images.githubusercontent.com/24503760/202043617-f48f6494-63f0-40a8-9d25-de6dd6738744.png)
![image](https://user-images.githubusercontent.com/24503760/202043494-a5200b6f-a095-4d07-9854-1d289786679b.png)

But I just realized afterward that the URL name filtering itself is already pretty strict and not allowing arbitrary characters to be inserted in, therefore I think excluding the filter could be done. Thus I agreed with the @parasharrajat and the rest of the team decision to not including the javascript.

Thank you so much and sorry for delayed response once again.

**Comment by kerupuksambel** [comment]
TL;DR : No need to do the rework javascript filtering, since it's already pretty safe to do the filtering with the URL parsing alone since they doesn't allow arbitrary characters on the URL already.

**Comment by parasharrajat** [comment]
Ok. Thanks. Unfortunately, we moved the issue to internal. @cead22 Do you think we can make this external if you haven't already worked on this?

There is a proposal above.
